### PR TITLE
First working version

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Only limited core functionality for now

I think tests should be extended

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+
+from setuptools import setup
+from catkin_pkg.python_setup import generate_distutils_setup
+
+d = generate_distutils_setup(
+    # #  don't do this unless you want a globally visible script
+    # scripts=['bin/myscript'],
+    packages=["ed_py"],
```
Comment (LoyVanBeek) [inline setup.py]: I still think this abbreviated name is inconsistent, not worth saving those 4 chars.

**Review by LoyVanBeek — DISMISSED** [review]

File: src/ed_py/entity.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+from .shape import shape_from_entity_info
+from .volume import Volume, volumes_from_entity_volumes_msg
+
+
+class Entity:
+    """Holds all data concerning entities"""
+
+    def __init__(
+        self,
+        identifier,
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Missing typehints

File: src/ed_py/entity.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+        # Check if the point is inside of the volume
+        return self._volumes[volume_id].contains(vector)
+
+    def entities_in_volume(self, entities, volume_id):
+        """
+        Filter the collection of entities down to only those in the given volume
+
+        :param entities: collection/sequence of entities
+        :type entities: List[Entity]
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Addition of type is inconsistent with docstrings above

File: src/ed_py/entity.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+    def __repr__(self):
+        return "Entity(id='{id}', type='{type}', frame={frame}, person_properties={pp})".format(
+            id=self.id, type=self.type, frame=self.pose, pp=self._person_properties
+        )
+
+
+class PersonProperties(object):
+    def __init__(
+        self,
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: no typehints

File: src/ed_py/entity.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        Container for several properties related to a person
+
+        :param name: the person's name. This is separate from the entity, which is unique while this doesn't have to be
+        :param age: Estimated age of the person
+        :param emotion: str indicating the emotion
+        :param gender: Predicted gender of the person
+        :param gender_confidence: Confidence of the classifier in the gender above.
+        :param pointing_pose: In which direction is the person pointing
+        :param posture: String with a value like 'sitting', 'laying', 'standing' etc.
+        :param reliability:  ?
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Yeah that's a no from me, no question marks in documentation svp

File: src/ed_py/entity.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        shape=shape,
+        volumes=volumes,
+        super_types=super_types,
+        last_update_time=last_update_time,
+    )
+
+    if e.type == "person":
+        try:
+            pp_dict = yaml.load(e.data)
+            del pp_dict["position"]
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Why is this necessary?

File: src/ed_py/shape.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        )
+
+
+class RightPrism(Shape):
+    """
+    Represents a right prism, i.e., a  prism in which the joining edges and faces are perpendicular to the base
+    faces. This is typical for the shapes resulting from the convex hull, z min and z max of EntityInfo
+    """
+
+    def __init__(self, convex_hull, z_min, z_max):
```
Comment (LarsJanssenTUe) [inline src/ed_py/shape.py]: Missing typehints

File: src/ed_py/shape.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    def _calc_size(self):
+        """
+        Calculate the rough size of a shape
+        >>> RightPrism([kdl.Vector(0, 0, 0), kdl.Vector(0, 1, 0), kdl.Vector(1, 1, 0), kdl.Vector(1, 0, 0)], z_min=0, z_max=1).size
+        1.0
+        >>> RightPrism([kdl.Vector(0, 0, 0), kdl.Vector(0, 2, 0), kdl.Vector(2, 2, 0), kdl.Vector(2, 0, 0)], z_min=0, z_max=2).size
+        8.0
+        """
+
+        # TODO: this only uses the bounding volume's size, which might not be accurate
```
Comment (LarsJanssenTUe) [inline src/ed_py/shape.py]: Add what should be used to this TODO, this is not clear to me

File: src/ed_py/volume.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+import tf2_ros
+
+# noinspection PyUnresolvedReferences
+import tf2_kdl
+
+
+class Volume:
+    """
+    Represents an volume of an entity
```
Comment (LarsJanssenTUe) [inline src/ed_py/volume.py]:
```suggestion
    Represents a volume of an entity
```

File: src/ed_py/volume.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    """Represents a box shaped volume"""
+
+    def __init__(self, min_corner, max_corner):
+        """
+        Constructor
+
+        Points are defined relative to the object they belong to
+
+        :param min_corner: PyKDL.Vector with the minimum bounding box corner
+        :param max_corner: PyKDL.Vector with the minimum bounding box corner
```
Comment (LarsJanssenTUe) [inline src/ed_py/volume.py]:
```suggestion
        :param max_corner: PyKDL.Vector with the maximum bounding box corner
```

File: src/ed_py/world_model.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            prefix = f"/{ns}/"
+
+        self._ed_simple_query_srv = self.create_service_client(f"{prefix}ed/simple_query", SimpleQuery)
+        self._ed_update_srv = self.create_service_client(f"{prefix}ed/update", UpdateSrv)
+
+        self._ed_configure_srv = self.create_service_client(f"{prefix}ed/configure", Configure)
+        self._ed_reset_srv = self.create_service_client(f"{prefix}ed/reset", Reset)
+
+    @staticmethod
+    def create_service_client(name: str, srv_type):
```
Comment (LarsJanssenTUe) [inline src/ed_py/world_model.py]: typehints, please make consistent in entire PR

File: src/ed_py/world_model.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+            return []
+
+        return entities
+
+    def get_closest_entity(self, center_point: VectorStamped, etype: str = "", radius: float = float("inf")) -> Entity:
+
+        entities = self.get_entities(center_point=center_point, etype=etype, radius=radius)
+
+        # HACK
+        entities = [e for e in entities if e.shape is not None and e.etype != ""]
```
Comment (LarsJanssenTUe) [inline src/ed_py/world_model.py]: Please specify what you would like it to be

File: src/ed_py/world_model.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+                f"ignore_z={ignore_z})\n{e}, "
+                f"{traceback.format_exc()}"
+            )
+            return []
+
+        return entities
+
+    def get_closest_entity(self, center_point: VectorStamped, etype: str = "", radius: float = float("inf")) -> Entity:
+
+        entities = self.get_entities(center_point=center_point, etype=etype, radius=radius)
```
Comment (LarsJanssenTUe) [inline src/ed_py/world_model.py]: What entities do you get when you leave etype the default value?

**Review by LarsJanssenTUe — CHANGES_REQUESTED** [review]

File: src/ed_py/entity.py (hunk 5 above)
Comment (MatthijsBurgh) [reply src/ed_py/entity.py]: But I don't know either

File: src/ed_py/entity.py (hunk 6 above)
Comment (MatthijsBurgh) [reply src/ed_py/entity.py]: Because these aren't accepted by the class. So you don't have to add all arguments separately.

File: src/ed_py/world_model.py (hunk 13 above)
Comment (MatthijsBurgh) [reply src/ed_py/world_model.py]: If no type is specified, all entities that satisfy the other constrains are returned

File: src/ed_py/world_model.py (hunk 12 above)
Comment (MatthijsBurgh) [reply src/ed_py/world_model.py]: This copied from the robot part. So it is unknown how hacky this is.

File: package.xml [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  <depend>ed_msgs</depend>
+
+  <exec_depend condition="$ROS_PYTHON_VERSION == 2">python-numpy</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-numpy</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 2">python-yaml</exec_depend>
+  <exec_depend condition="$ROS_PYTHON_VERSION == 3">python3-yaml</exec_depend>
+  <exec_depend>pykdl_ros</exec_depend>
+  <exec_depend>python_orocos_kdl</exec_depend>
+  <exec_depend>rospy</exec_depend>
+  <exec_depend>std_msgs</exec_depend> <!--[TEMP] Waiting for change in ed_msgs -->
```
Comment (LarsJanssenTUe) [inline package.xml:28]: Please make the relevant changes to remove this temporary dependency

File: setup.py (hunk 1 above)
Comment (LarsJanssenTUe) [reply setup.py]: I also agree, our codebase can handle the 4 extra chars.

File: src/ed_py/entity.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    def __init__(
+        self,
+        identifier: str,
+        object_type: str,
+        frame_id: str,
+        pose: FrameStamped,
+        shape: Shape,
+        volumes: Mapping[str, Volume],
+        super_types: List[str],
+        last_update_time,
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Missing typehint

File: src/ed_py/entity.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+        """
+        return super_type in self.super_types
+
+    @property
+    def pose(self) -> FrameStamped:
+        """Returns the pose of the Entity as a FrameStamped"""
+        return FrameStamped(self._pose, self.last_update_time, self.frame_id)
+
+    @pose.setter
+    def pose(self, pose):
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: missing typehint

File: src/ed_py/entity.py [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+            return self._person_properties
+        else:
+            rospy.logwarn("{} is not a person".format(self))
+            return None
+
+    @person_properties.setter
+    def person_properties(self, value: PersonProperties):
+        self._person_properties = value
+
+    def __repr__(self):
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: missing output type

File: src/ed_py/entity.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+    def name(self) -> str:
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        rospy.loginfo("Changing {}'s name to {}".format(self._parent_entity.id, value))
+        self._name = value
+
+    def __repr__(self):
+        return "PersonProperties(age='{age}', gender='{g}', gender_confidence={gc}, shirt_colors={sc})".format(
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: missing output type

File: src/ed_py/entity.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+    Converts ed_msgs.msg.EntityInfo to an Entity
+
+    :param e: ed_msgs.msg.EntityInfo
+    :return: Entity
+    """
+    assert isinstance(e, EntityInfo)
+    identifier = e.id
+    object_type = e.type
+    frame_id = "map"  # ED has all poses in map
+    # ToDo: Change pose in msg definition to PoseStamped
```
Comment (LarsJanssenTUe) [inline src/ed_py/entity.py]: Can/should this be done now?

File: src/ed_py/shape.py [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+    """Represents shape properties"""
+
+    def __init__(self):
+        pass
+
+    @property
+    def convex_hull(self) -> List[kdl.Vector]:
+        return self._calc_convex_hull()
+
+    def _calc_convex_hull(self):
```
Comment (LarsJanssenTUe) [inline src/ed_py/shape.py]: missing output type

File: src/ed_py/shape.py [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+    def _calc_size(self) -> float:
+        """
+        Calculate the rough size of a shape
+        >>> RightPrism([kdl.Vector(0, 0, 0), kdl.Vector(0, 1, 0), kdl.Vector(1, 1, 0), kdl.Vector(1, 0, 0)], z_min=0, z_max=1).size
+        1.0
+        >>> RightPrism([kdl.Vector(0, 0, 0), kdl.Vector(0, 2, 0), kdl.Vector(2, 2, 0), kdl.Vector(2, 0, 0)], z_min=0, z_max=2).size
+        8.0
+        """
+
+        # TODO: this only uses the bounding box's size, which might not be accurate
```
Comment (LarsJanssenTUe) [inline src/ed_py/shape.py]: Okay so what is missing in this approach? Why might this not be accurate?

File: src/ed_py/volume.py [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+            if (
+                min_corner.x() <= point.x() <= max_corner.x()
+                and min_corner.y() <= point.y() <= max_corner.y()
+                and min_corner.z() <= point.z() <= max_corner.z()
+            ):
+                return True
+
+        return False
+
+    def __repr__(self):
```
Comment (LarsJanssenTUe) [inline src/ed_py/volume.py]: missing output type

File: src/ed_py/volume.py [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+        for min_corner, max_corner in zip(self._min_corners, self._max_corners):
+            description += "min_corner={}, max_corner={}\n".format(min_corner, max_corner)
+
+        return description
+
+
+class OffsetVolume(Volume):
+    """Represents a volume with a certain offset from the convex hull of the entity"""
+
+    def __init__(self, offset):
```
Comment (LarsJanssenTUe) [inline src/ed_py/volume.py]: missing typehint

File: src/ed_py/volume.py [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+        """Constructor
+
+        :param offset: Offset [m]
+        """
+        super(OffsetVolume, self).__init__()
+        self._offset = offset
+
+    def __repr__(self):
```
Comment (LarsJanssenTUe) [inline src/ed_py/volume.py]: missing output type

File: src/ed_py/world_model.py [hunk 25]
```diff
@@ -0,0 +1,255 @@
...
+            prefix = f"/{ns}/"
+
+        self._ed_simple_query_srv = self.create_service_client(f"{prefix}ed/simple_query", SimpleQuery)
+        self._ed_update_srv = self.create_service_client(f"{prefix}ed/update", UpdateSrv)
+
+        self._ed_configure_srv = self.create_service_client(f"{prefix}ed/configure", Configure)
+        self._ed_reset_srv = self.create_service_client(f"{prefix}ed/reset", Reset)
+
+    @staticmethod
+    def create_service_client(name: str, srv_type: type):
```
Comment (LarsJanssenTUe) [inline src/ed_py/world_model.py]: missing output type

File: src/ed_py/world_model.py (hunk 12 above)
Comment (LarsJanssenTUe) [reply src/ed_py/world_model.py]: Then please at least state that it stems from there

**Review by LarsJanssenTUe — CHANGES_REQUESTED** [review]

File: package.xml (hunk 14 above)
Comment (MatthijsBurgh) [reply package.xml:28]: Changing this will create too much works on other repo's. So to limit this, we should leave this for now.

File: src/ed_py/entity.py (hunk 17 above)
Comment (MatthijsBurgh) [reply src/ed_py/entity.py]: `__repr__` always should return a string.

File: src/ed_py/entity.py (hunk 4 above)
Comment (MatthijsBurgh) [reply src/ed_py/entity.py]: Not needed, not possible, not clever

File: src/ed_py/entity.py (hunk 19 above)
Comment (MatthijsBurgh) [reply src/ed_py/entity.py]: Related to review in package.xml. This would require a lot of changes in other repos

**Review by LarsJanssenTUe — APPROVED** [review]
